src/sawtoothMain/kotlin/lumberjack/sawtooth/layout/pattern/PatternLayout.kt

```diff
@@ -0,0 +1,99 @@
+        private const val TOKEN_MODIFIER_PATTERN = "\\{(\\w+)\\}"
+
+        /**
+         * Regex is notoriously hellish, so let's run through what this does
+         * This is a regex string to match a token - that is, a percent sign '%' followed by one or more word characters, optionally enclosed with brackets
+         * The regex functions like so:
+         * - `(?<=[^%]|^)` Requires the first character to either be anything *other* than a percent sign, *or* have it be the start of the string. Does not match that character.
+         * - `%` Match a single percent sign
+         * - `(?:(\w+)|(?:\[(\w+)\]))` Use a non-capturing group to select one of two options - either match one or more word characters, or match an opening bracket, one or more word characters, and then a closing bracket
+         * - `((?:$TOKEN_MODIFIER_PATTERN)*)` Use a non-capturing group to match the token modifier pattern zero or more times, and capture that whole result
+         */
+        private val tokenRegex = "(?<=[^%]|^)%(?:(\\w+)|(?:\\[(\\w+)\\]))((?:$TOKEN_MODIFIER_PATTERN)*)".toRegex()
+
+        /** Because regex can't capture a group multiple times, we need a separate regex to find all matches in the modifier match */
+        private val tokenModifierRegex = TOKEN_MODIFIER_PATTERN.toRegex()
+
+        fun fromPattern(
+            pattern: String,
+            componentFactory: PatternComponentFactory = DefaultPatternComponentFactory
+        ): PatternLayout {
+            val components = ArrayList<PatternComponent>(1)
+
+            var start = 0
+            //Find our first match in the pattern
+            var tokenMatch = tokenRegex.find(pattern, start)
+
+            //While we have a valid match
+            while (tokenMatch != null) {
+                //If the match is further into the string than the starting position, we have some text to write as-is
+                if (tokenMatch.range.first > start) {
+                    components.add(LiteralComponent(pattern.substring(start, tokenMatch.range.first)))
+                }
+
+                //Get the token (remember: groupValues[0] is the whole matched string)
+                //If groupValues[1] is empty, it means we matched a sequence with brackets around it, so get the second group value
+                val token = tokenMatch.groupValues[1].takeUnless(String::isEmpty) ?: tokenMatch.groupValues[2]
+
+                val modifiers = ArrayList<String>(0)
+                val patternComponent = componentFactory.fromToken(token, modifiers)
+                if (patternComponent != null) {
+                    //If we have an initializer, invoke it and add it to the list of components
+                    //The list of modifiers is obtained by finding all valid matches of tokenModifierRegex on the third group value
+                    tokenModifierRegex.findAll(tokenMatch.groupValues[3]).mapTo(modifiers) { result -> result.groupValues[1] }
```

Comment:
Actually now I can see that this can be problematic. Is there a way to detect if there are modifiers, and to initialize them, before creating the `patternComponent`?

Reply:
Fixed with 32a888f